### PR TITLE
Un-revert Jinja2 fix

### PR DESCRIPTION
**Pull request by babolivier**
* Reverts https://github.com/matrix-org/synapse/pull/12296
* Pins `Jinja2>=3.0`

Fixes https://github.com/matrix-org/synapse/issues/12294

### CODE REVIEW

File: changelog.d/12313.misc [hunk 1]
```diff
@@ -0,0 +1 @@
+Correctly import `Markup` from `MarkupSafe` instead of `Jinja2`.
```
Comment (richvdh) [inline changelog.d/12313.misc:1]: could you expand on this? Why are we doing this?

File: synapse/python_dependencies.py [hunk 2]
```diff
@@ -74,8 +74,7 @@
     # Note: 21.1.0 broke `/sync`, see #9936
     "attrs>=19.2.0,!=21.1.0",
     "netaddr>=0.7.18",
-    # Jinja2 3.1.0 removes the deprecated jinja2.Markup class, which we rely on.
-    "Jinja2<3.1.0",
+    "Jinja2~=3.0",
```
Comment (richvdh) [inline synapse/python_dependencies.py:77]: why do we need 3.x? a comment would be good.

File: changelog.d/12313.misc [hunk 3]
```diff
@@ -0,0 +1 @@
+Import `Markup` from `MarkupSafe` as importing it from `Jinja2` has been deprecated for some time and recently removed.
```
Comment (richvdh) [inline changelog.d/12313.misc:1]: How about simply:

```suggestion
Fix compatibility with the recently-released Jinja 3.1?
```

File: synapse/python_dependencies.py [hunk 4]
```diff
@@ -74,8 +74,8 @@
     # Note: 21.1.0 broke `/sync`, see #9936
     "attrs>=19.2.0,!=21.1.0",
     "netaddr>=0.7.18",
-    # Jinja2 3.1.0 removes the deprecated jinja2.Markup class, which we rely on.
-    "Jinja2<3.1.0",
+    # Jinja 2.x is incompatible with recent versions of MarkupSafe, on which it depends.
+    "Jinja2~=3.0",
```
Comment (richvdh) [inline synapse/python_dependencies.py:78]: The `olddeps` logic ([here](https://github.com/matrix-org/synapse/blob/develop/tox.ini#L125)) doesn't handle `~=`, so you'll either need to fix that, or change this to a simple `>=`.

Comment (babolivier) [reply synapse/python_dependencies.py:80]:
```suggestion
    "Jinja2>=3.0",
```

**Review by richvdh — APPROVED** [review]
thanks!

**Pull request merged**